### PR TITLE
ÖAW ACDH updated with separated EntityDescriptors.

### PR DESCRIPTION
**Pull request by dpancic**

### CODE REVIEW

**Comment by andmor-** [comment]
Please check QA results in https://clarin-eric.github.io/SPF-SPs-metadata/page/master_qa_report.html

-> Can English ('en') descriptions and displayname lengths be fixed?

**Comment by andmor-** [comment]
The new metadata as been pushed to our IdP see:
https://infra.clarin.eu/aai/md_about_spf_sps.xml

Can you please test login each of the new entities using the CLARIN homeless IdP?
Alternatively you can provide me the login URLs for each of the endpoints.

Note: the new metadata is still on the CLARIN IdP only, i.e. it is still not pushed to the various identity federations of the CLARIN SPF.

**Comment by dpancic** [comment]
thank you,

In our shibboleth configuration we use both
aconet-interfederation and CLARIN IdP metadata (.
https://infra.clarin.eu/aai/prod_md_about_clarin_erics_idp.xml and
http://idp.mpi.nl/metadata/homeless-clarin-metadata.xml)
On our idp list is only clarin.eu site account and when it is selected it
returnes:

eu.unicore.samly2.exceptions.SAMLRequesterException:
AssertionConsumerServiceURL in request (
https://teach.dariah.eu/Shibboleth.sso/SAML2/POST) is not among trusted
endpoints of the issuer.Caused by:
eu.unicore.samly2.exceptions.SAMLRequesterException:
AssertionConsumerServiceURL in request (
https://teach.dariah.eu/Shibboleth.sso/SAML2/POST) is not among trusted
endpoints of the issuer.

One of our services is #dariahTeach and it can be tested:
https://teach.dariah.eu/login/index.php

2018-01-16 11:38 GMT+01:00 andmore <notifications@github.com>:

**Comment by andmor-** [comment]
I can confirm the problem but by now I don't have a clear idea about the cause.

Is your SP metadata completely in sync with the metadata supplied in this pull request?

You mention you use `http://idp.mpi.nl/metadata/homeless-clarin-metadata.xml` in the SP but this URL is not valid anymore. The only IdP metadata we use is the one coming from the first URL you posted, so you can remove the second.

Can you supply me the SP logs of a login attempt up to the error?

**Comment by andmor-** [comment]
The problem seems to be related with the `entityID` declared by your discovery service.

When I go to: `https://teach.dariah.eu/login/index.php` and select `Shibboleth login`. Your discovery service shows me the following URL on the address bar
`https://acdh.oeaw.ac.at/shibboleth-ds/index.html?entityID=https%3A%2F%2Fclarin.oeaw.ac.at%2Fshibboleth&return=https%3A%2F2Fteach.dariah.eu%2FShibboleth.sso%2FLogin%3FSAMLDS%3D1%26target%3Dss%253Amem%253A0bafc46820c2ed0b2b13f609f5bb3692c04233852a1f9ea114cc2a92de43a2ef`

Note the `entityID=https%3A%2F%2Fclarin.oeaw.ac.at%2Fshibboleth` part.

Now if I change the `entityID` of `https://teach.dariah.eu/shibboleth` in your metadata to: `entityID=https://clarin.oeaw.ac.at/shibboleth` I am able to login without problems!

Can your discovery service dinamically use different `entityID`'s depending on which SP requested the authentication?

**Comment by andmor-** [comment]
screen shot 2018-01-22 at 14 08 55

**Comment by andmor-** [comment]
screen shot 2018-01-22 at 14 16 41

**Comment by dpancic** [comment]
We use The Embedded Discovery Service (EDS)  packaged for RHEL 7.


2018-01-22 14:18 GMT+01:00 andmore <notifications@github.com>:

**Comment by andmor-** [comment]
Do you have `entityID` declared in the SP's default session initiator (in shibboleth2.xml)?
Can share this file with us? (via encrypted email if you prefer)

**Comment by dpancic** [comment]
Our settings is:
```

<ApplicationDefaults entityID="https://clarin.oeaw.ac.at/shibboleth"
                               REMOTE_USER="eppn persistent-id targeted-id">

<SSO discoveryProtocol="SAMLDS" discoveryURL="https://acdh.oeaw.ac.at/shibboleth-ds/index.html">
                SAML2 SAML1
</SSO>

```
        <MetadataProvider xmlns:xi="http://www.w3.org/2001/XInclude" type="Chaining">
            
            <MetadataProvider type="XML"
                              uri="http://eduid.at/md/aconet-interfed.xml"
                              backingFilePath="aconet-interfed-metadata.xml"
                              reloadInterval="7200">
                <MetadataFilter type="Signature" certificate="aconet-metadata-signing.crt"/>
                <MetadataFilter type="RequireValidUntil" maxValidityInterval="2419200"/>
                <MetadataFilter type="EntityRoleWhiteList">
                    <RetainedRole>md:IDPSSODescriptor</RetainedRole>
                    <RetainedRole>md:AttributeAuthorityDescriptor</RetainedRole>
                </MetadataFilter>
            </MetadataProvider>

            
            <MetadataProvider type="XML"
                              uri="https://infra.clarin.eu/aai/prod_md_about_spf_idps.xml"
                              backingFilePath="prod_md_about_spf_idps.xml"
                              reloadInterval="7200">
                <MetadataFilter type="Blacklist" xmlns="urn:mace:shibboleth:2.0:native:sp:config">
                    <Exclude>https://openidp.aco.net/saml</Exclude>
                </MetadataFilter>
            </MetadataProvider>

            
            <MetadataProvider type="XML"
                              uri="https://infra.clarin.eu/aai/prod_md_about_clarin_erics_idp.xml"
                              backingFilePath="prod_md_about_clarin_erics_idp.xml"
                              reloadInterval="7200"/>

              <MetadataProvider type="XML"
                              uri="http://idp.mpi.nl/metadata/homeless-clarin-metadata.xml"
                              backingFilePath="clarin-eu-metadata.xml"
                              reloadInterval="2419200"/>

**Comment by andmor-** [comment]
You only have one application `entityID` defined, and you need to have one for each application interfacing with the SP.

What you need is to define one `ApplicationOverride` block for each of the applications. See: https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApplicationOverride
This is usually found but commented out in the default settings.xml file from shibboleth.

In this case, something like this (replace {{XXX}} for your various entity Ids):
```
<RequestMapper type="XML">
  <RequestMap>
    <Host name="teach.dariah.eu"
            authType="shibboleth"
            requireSession="true"
            redirectToSSL="443"
            applicationId="teach_sp"
            exportAssertion="false">
        <Path name="/" />
     </Host>
  </RequestMap>
...
  <Host name="{{SP2_HOST}}"
            authType="shibboleth"
            requireSession="true"
            redirectToSSL="443"
            applicationId="sp2"
            exportAssertion="false">
        <Path name="/" />
     </Host>
  </RequestMap>
...
<ApplicationDefaults entityID="https://clarin.oeaw.ac.at/shibboleth"
                               REMOTE_USER="eppn persistent-id targeted-id">
...
        <ApplicationOverride id="teach_sp" entityID="https://teach.dariah.eu/shibboleth" />
...
        <ApplicationOverride id="sp2" entityID="{{SP2_ENTITY_ID}}" />
...
</ApplicationDefaults>
```


Also this element should be removed:
```
          <MetadataProvider type="XML"
                          uri="http://idp.mpi.nl/metadata/homeless-clarin-metadata.xml"
                          backingFilePath="clarin-eu-metadata.xml"
                          reloadInterval="2419200"/>
```

**Comment by dpancic** [comment]
After updating Shibboleth config with

<ApplicationOverride id="teach_sp" entityID="
> https://teach.dariah.eu/shibboleth" />


and Apache config with:

ShibRequestSetting applicationId teach_sp


Following error appeard:

ERROR
> SAML service got an invalid request.
> If you are a user then you can be sure that the web application you was
> using previously is either misconfigured or buggy.
> If you are an administrator or developer, here the details of the error
> follows:
> eu.unicore.samly2.exceptions.SAMLRequesterException:
> AssertionConsumerServiceURL in request (
> https://teach.dariah.eu/Shibboleth.sso/SAML2/POST) is not among trusted
> endpoints of the issuer.
> Caused by: eu.unicore.samly2.exceptions.SAMLRequesterException:
> AssertionConsumerServiceURL in request (
> https://teach.dariah.eu/Shibboleth.sso/SAML2/POST) is not among trusted
> endpoints of the issuer.


2018-01-26 14:07 GMT+01:00 andmore <notifications@github.com>:

**Comment by andmor-** [comment]
Looks like apache is not applying the mapping. It's the same error.
Is there some error in the apache logs?
Are you applying `ShibRequestSetting applicationId teach_sp` globally or at the Location level?
Do you have the same error if you try to do the mapping in shibboleth2.xml?

Can you share you apache config?

**Comment by dpancic** [comment]
Apache log does not contain errors.
ShibRequestSetting applicationId teach_sp is applied at  the Location level.

 <Location /auth/shibboleth/index.php>
                AuthType shibboleth
                ShibRequestSetting requireSession 1
                ShibRequestSetting applicationId teach_sp
                Require shib-session
                RequestHeader set eppn %{eppn}e
</Location>


When I tried to do mapping trough shibboleth2.xml it does not
return shibboleth-ds and produces error 500.



2018-01-30 10:45 GMT+01:00 andmore <notifications@github.com>:

**Comment by andmor-** [comment]
Can you try it for the webapplication root, as in https://wiki.library.ucsf.edu/display/IAM/Steps+to+configure+multiple+context+in+Shib+SP?

In your current configuration (not applied to the root), I don't think this location block is triggered for `https://teach.dariah.eu/Shibboleth.sso/Login` and it must be as outlined in the documentation:
```
## Note, if you can't apply this globally, you need to make sure
## you apply the setting to both the content/resources *and* to
## the appropriate handler path (generally /Shibboleth.sso).
```
Also to all content/resources served via `https://teach.dariah.eu/`

**Comment by dpancic** [comment]
Problem with NativeSPApplicationOverride is solved by creating the SP metadata by hand and  writing them in a file which is called from ApplicationOverride part of Shibboleth configuration.

**Comment by vronk** [comment]
Here some more details on the resolution:

1. The XML describing the individual entities (the whole md:EntityDescriptor XML element describing the SP) is saved in individual files under: `/etc/shibboleth/metadata/{SP_domain}.xml`
2. In `/etc/shibboleth/shibboleth2.xml`  added
     `<MetadataProvider type="XML" file="metadata/{SP_domain}.xml"/>`
        to the `<MetadataProvider type="Chaining">` section.
     and added:
       `<ApplicationOverride id="{SP_domain}" entityID="{SP_entityID}"/>` 
        to the `<ApplicationDefaults> section.`
3.  in `/etc/httpd/conf.d/sites-enabled/{SP_domain}.conf` added `ShibRequestSetting` directive to every `<Location>`  in the HTTPS virtual host config:
        `<Location />
            ShibRequestSetting applicationId {SP_domain}
        </Location>`

4. Restart the Shibboleth and reload the Apache
    `service httpd reload && service shibd restart`

Remark. The {SP_domain} used above is only a convention. In fact it can be any string, just consistently used in all places. Anyway using SP's domain makes it easier to quickly identify to which SP a given config part relates.

### Notes on how and why it works:
To make everything work SPs Shibboleth must load both IdPs metadata and SPs metadata.
Shibboleth stores both of them together in RAM and simply matches on IdP/SP entityID value.
The source of all these metadata is configured by the `<MetadataProvider>` section of the `/etc/shibboleth/shibboleth2.xml`.
There is one exception to that - the Shibboleth automatically populates its metadata database with a metadata entry describing IdP/SP as configured in the `<ApplicationDefaults>` section of the `/etc/shibboleth/shibboleth2.xml.
Anyway metadata are automatically generated only for the default IdP/SP set up in the Shibboleth config and all additional IdPs/SPs metadata must be explicitly imported using a proper entry in the <MetadataProvider> section of the /etc/shibboleth/shibboleth2.xml (see the "simply make it done" instruction above.

The good thing about the Shibboleth approach is we can use one global metadata database for all SPs (therefore save a lot of memory).
Of course that's possible only until all SPs are coupled with exactly the same IdPs (but fortunately that's our case).
If we would like to use different IdPs with different SPs, we would simply need to include the `<MetadataProvider>` section in every SP's `<ApplicationOverride>` section.

**Comment by andmor-** [comment]
Thank you very much for the detailed explanation. Hopefully this can help people in the future if they run into this issue.

**Comment by peter-** [comment]
Note that chosing [Log in](https://teach.dariah.eu/login/index.php) and "Shibboleth login" at teach.dariah.eu links to https://clarin.oeaw.ac.at/moodledariahteach/auth/shibboleth/index.php and picking `https://idm.clarin.eu` as the IDP in the [EDS there](https://acdh.oeaw.ac.at/shibboleth-ds/index.html) still leads to that same error, i.e. the issue still exists!

The SAML authentication request contains the entityID `https://clarin.oeaw.ac.at/shibboleth` as Issuer, and `https://teach.dariah.eu/Shibboleth.sso/SAML2/POST` as ACS URL, and the CLARIN IDP b0rks at that (`AssertionConsumerServiceURL in request (https://teach.dariah.eu/Shibboleth.sso/SAML2/POST) is not among trusted endpoints of the issuer.`). There are several aspects to this:

* I'm not surprised that entityID is the Issuer of the authn request, as the request will be generated by the Shib SP running at the vhost https://clarin.oeaw.ac.at/ and even after applying any `ApplicationOverride`s that vhost would still issue auth requests as "itself", not as teach.dariah.eu. So I'd first question why for logging in to teach.dariah.eu one is redirected to https://clarin.oeaw.ac.at/moodledariahteach/auth/shibboleth/index.php -- i.e., I think that's the underlying mistake, Moodle and its login need to be accessible on the correct vhost so that the Shib SP can issue authn requests with the correct entityID (as per the application mapping per vhost).
* Note that the eduID.at-curated metadata for `https://clarin.oeaw.ac.at/shibboleth` actually does contain the ACS URL for `https://teach.dariah.eu/Shibboleth.sso/SAML2/POST`, so if the IDP in question
  1. had correct/complete metadata for your services available, and
  2. would support multiple ACS URLs of the same binding for a single entityID, this would even work, though not using the separated entityIDs (but the old approach of a single entityID with protocol endpoints for every vhost).

Some other remarks to this thread:
* This has nothing to do with IDP Discovery, as [speculated](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-359415552), but is the result of all of the bullets above (web server misconfiguiration plus incomplete metadata and/or buggy IDP implementation).
* This being on Apache httpd the SP's `RequestMap` should not be used, contrary to [this suggestion](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-360779924), instead use native httpd directives.
* Since you want the overrides to apply to each vhost [this mapping](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-361863217) is insufficient, you'll want to place it a `<Location />` directive covering the whole vhost, as [mentioned](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-361923706) by André.
* The [resolution](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-363683397) and [explanation](https://github.com/clarin-eric/SPF-SPs-metadata/pull/2#issuecomment-364088299) do not make much sense to me:
Ad (1): The SP only needs to have an `ApplicationOverride` element for each logical "application" plus the `entityID` to use. See the first example in the [Shibboleth documentation](https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApplicationOverride#NativeSPApplicationOverride-CommonExampleScenarios). And httpd only needs to know where to apply those "applications".
Nothing in the SP has anything to do with SAML Metadata *about* the SP!
Ad (2): The SP does not need its own metadata, so the first step here is nonsensical. The second step is just what the Shib documentation says.
Ad (3): Adding the application once to the whole vhost should suffice (not *every* Location directive)
Step 4 is correct.

> To make everything work SPs Shibboleth must load both IdPs metadata and SPs metadata.

No. I think pretty much all statements about metadata in the section "how and why it works" are either wrong (claimed need for the SP to ingest SP metadata) or irrelevant (internal implementation details).

To sum it up: Issue still exists, likely reason is server misconfig, even after applying the SP overrides.

**Comment by zozlak** [comment]
> Nothing in the SP has anything to do with SAML Metadata about the SP!

Point 1-3 are not about SAML protocol but about *how to convince the Shibboleth to generate requests with the right SP entityID/metadata in the "multiples SPs on one Shibboleth instance" setup*:

* The `ShibRequestSetting applicationId {SP_domain}` directive coupled with `<ApplicationOverride id="{SP_domain}" entityID="{SP_entityID}"/>` didn't do a job for us. We were still getting errors from IdPs suggesting our SP is set up wrong.
Also when looking at the `https://{SPdomain}/Shibboleth.sso/Metadata` the Shibboleth provided metadata of the default SP and not the correct one (which btw is in line with the Shibboleth documentation).
So my guess was (and your observations seems to confirm that) that the Shibboleth used wrong SP (basically always served the default SP entityID/metadata) when it was unable to match the `entityID` set up in the `<ApplicationOverride>` with any entry in its metadata database.
* Including SP's metadata into Shibboleth config automagically solved the issue (login requests started to be accepted by IdPs and the correct SP metadata were displayed at `https://{SPdomain}/Shibboleth.sso/Metadata`).

> Adding the application once to the whole vhost should suffice (not every Location directive)
Step 4 is correct.

From the tests we performed it wasn't. It was sufficient if the whole vhost shared the same Shibboleth config but in a setup like:

* whole vhost set up in a *initialize a session if it exists, don't complain if it doesn't* way
* particular *login location* set up in the *enforce shibboleth login* way

it was required to repeat the `ShibRequestSetting applicationId {SP_domain}` directive for the *login location* to make it work.

**Comment by zozlak** [comment]
I also think it will be useful to distinct between:

* ACDH SPs registration in federations and ACDH Shibboleth + proxy Apache config
  This is something we are focused at in this topic.
  When in the previous post I say something works (or doesn't), I mean I can (can't) properly log in/log out by interacting only with these two apps.
* ACDH service served under ACDH SPs domains
  This is a separate issue. Many of them can be set up poorly or wrong (e.g. generate login requests with wrong return URL like old CCV or redirect to a login endpoint of another SP). Of course it's also important and we should fix it but it has nothing to do with the way ACDH SPs are being registered in federations nor with the way ACDH Shibboleth is set up.
